app/components/Reactions/Reaction.tsx

```diff
@@ -0,0 +1,173 @@
+import { observer } from "mobx-react";
+import { darken, lighten } from "polished";
+import React from "react";
+import { useTranslation } from "react-i18next";
+import styled, { css } from "styled-components";
+import { s } from "@shared/styles";
+import type { Reaction as TReaction } from "@shared/types";
+import { getEmojiId } from "@shared/utils/emoji";
+import User from "~/models/User";
+import { Emoji } from "~/components/Emoji";
+import Flex from "~/components/Flex";
+import NudeButton from "~/components/NudeButton";
+import Text from "~/components/Text";
+import Tooltip from "~/components/Tooltip";
+import useCurrentUser from "~/hooks/useCurrentUser";
+import { hover } from "~/styles";
+import { ReactionData } from "~/types";
+
+const MaxUsernamesInTooltip = 3;
+
+type Props = {
+  /** The reaction data - has the emoji & active user ids for this reaction. */
+  reaction: TReaction;
+  /** Data of the users who have reacted using this emoji. */
+  reactedUsers: ReactionData["user"][];
+  /** Callback when the user intends to add the reaction. */
+  onAddReaction: (emoji: string) => Promise<void>;
+  /** Callback when the user intends to remove the reaction. */
+  onRemoveReaction: (emoji: string) => Promise<void>;
+};
+
+const useTooltipContent = ({
+  reactedUsers,
+  currUser,
+  emoji,
+  active,
+}: {
+  reactedUsers: Props["reactedUsers"];
+  currUser: User;
+  emoji: string;
+  active: boolean;
+}) => {
+  const { t } = useTranslation();
+
+  if (!reactedUsers.length) {
+    return;
+  }
+
+  const usernames: string[] = [];
+
+  if (active) {
+    usernames.push(t("You"));
+  }
+
+  const otherUsernames = reactedUsers
+    .filter((user) => user.id !== currUser.id)
+    .map((user) => user.name);
+
+  usernames.push(
+    ...otherUsernames.slice(
+      0,
+      active ? MaxUsernamesInTooltip - 1 : MaxUsernamesInTooltip
+    )
+  );
+
+  const diff = reactedUsers.length - usernames.length;
+  if (diff > 0) {
+    usernames.push(`${diff} ` + (diff === 1 ? t("other") : t("others")));
+  }
+
+  const joinedUsernames = usernames.reduce((content, name, idx) => {
+    if (idx === 0) {
+      return name;
+    } else if (idx === usernames.length - 1) {
+      return `${content} ${t("and")} ${name}`;
+    } else {
+      return `${content}, ${name}`;
+    }
+  }, "");
+
+  return `${joinedUsernames} ${t("reacted with")} :${getEmojiId(emoji)}:`;
```

Comment:
Hard to provide exact feedback on how to change this, but we can't construct translated sentences from parts like this as every language has different grammatical structures. It needs to be done using variables in the translation as much as possible, e.g.

`t("{{ usernames }} reacted with {{ emoji }}")`

It's okay to have a couple of variants

Reply:
We could end up with a few variants like
- `t("You reacted with {{ emoji }}", { emoji })`
- `t("{{ username }} reacted with {{ emoji }}", { username, emoji })`
- `t("You and {{ username }} reacted with {{ emoji }}", { username, emoji })`
- `t("{{ first_username }} and {{ second_username }} reacted with {{ emoji }}", { first_username, second_username, emoji })`
- a couple more for three users
- and finally something like `t("You, {{ usernames }}, and {{ rem_count }} others reacted with {{ emoji }}", { usernames,  rem_count, emoji }`

I'm not really sure if this is the right way to handle the `and` conjunction. Should it be added to the translated string like above?

Reply:
"You" can be considered a username, so the `t("You reacted with {{ emoji }}", { emoji })` variant isn't needed. I think that leaves us with 4 variants:

- `t("{{ username }} reacted with {{ emoji }}", { username, emoji })`
- `t("{{ first_username }} and {{ second_username }} reacted with {{ emoji }}", { first_username, second_username, emoji })`
- `t("{{ username }} and {{ count }} others reacted with {{ emoji }}", { username,  count, emoji })`

Note it's important for the variable to be `count` on that last one for the i18n lib to generate the plural and singular variations

Reply:
Oh, I was thinking that we need to translate "You" as well since it's not a username and more of a text that can be localised. But, I suppose the same can be argued for the actual usernames too - can be localised but not advisable to expose PII.

Reply:
Localization happens at build time, so there's no exposing of PII. Take a look at this for an example of what I mean:

https://github.com/outline/outline/blob/b6fc8fb4b111cccd3db7e6d1f7f678281470cf3e/app/scenes/Settings/components/FileOperationListItem.tsx#L117-L122